### PR TITLE
feat(credit_note): Add estimate service and API end points

### PR DESCRIPTION
**Pull request by vincent-pochet**
## Context

Today it's hard to compute amounts for credit note creation when coupons or multiple tax rates are applied.
The front end and API users have to compute the tax and coupon pro-rata, credit and refund amounts by themselves hopping that the result will match the amounts computed on the backend when validating the validation.

## Description

In order to make it easier to create a credit note, this PR adds a new API end point and GraphQL resolver to retrieve creditable/refundable amounts based on items amount.

It exposes:
- A new `GET /api/v1/credit_notes` route
- A new `creditNoteEstimate` resolver

### CODE REVIEW

File: spec/serializers/v1/credit_notes/estimate_serializer_spec.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+  it 'serializes the object' do
+    result = JSON.parse(serializer.to_json)
+
+    aggregate_failures do
+      expect(result[root_name]['lago_invoice_id']).to eq(estimated_credit_note.invoice_id)
+      expect(result[root_name]['invoice_number']).to eq(estimated_credit_note.invoice.number)
+      expect(result[root_name]['currency']).to eq(estimated_credit_note.currency)
+      expect(result[root_name]['taxes_amount_cents']).to eq(estimated_credit_note.taxes_amount_cents)
+      expect(result[root_name]['sub_total_excluding_taxes_amount_cents']).to eq(estimated_credit_note.sub_total_excluding_taxes_amount_cents)
```
Comment (jdenquin) [inline spec/serializers/v1/credit_notes/estimate_serializer_spec.rb:28]: Layout/LineLength: Line is too long. [141/120] (https://rubystyle.guide#max-line-length)

File: spec/serializers/v1/credit_notes/estimate_serializer_spec.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+      expect(result[root_name]['max_creditable_amount_cents']).to eq(estimated_credit_note.credit_amount_cents)
+      expect(result[root_name]['coupons_adjustment_amount_cents']).to eq(estimated_credit_note.coupons_adjustment_amount_cents)
```
Comment (jdenquin) [inline spec/serializers/v1/credit_notes/estimate_serializer_spec.rb:30]: Layout/LineLength: Line is too long. [127/120] (https://rubystyle.guide#max-line-length)

File: app/services/credit_notes/estimate_service.rb [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+      super
+    end
+
+    def call
+      return result.not_found_failure!(resource: 'invoice') unless invoice
+      return result.forbidden_failure! unless License.premium?
+      return result.not_allowed_failure!(code: 'invalid_type_or_status') unless valid_type_or_status?
+
+      credit_note = CreditNote.new(
```
Comment (lovrocolic) [inline app/services/credit_notes/estimate_service.rb:17]: It seems to me that `credit_note` could be an instance variable in this service based on context and that we can avoid passing it in many methods as a parameter. wdyt?

**Review by lovrocolic — APPROVED** [review]

File: spec/scenarios/credit_note_spec.rb [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    travel_to(DateTime.new(2023, 10, 23)) do
+      update_invoice(invoice, payment_status: :succeeded)
+
+      estimate_credit_note(
+        invoice_id: invoice.id,
+        items: [
+          {
+            fee_id: fee2.id,
+            amount_cents: 26_260,
+          }
```
Comment (jdenquin) [inline spec/scenarios/credit_note_spec.rb:101]: Style/TrailingCommaInArrayLiteral: Put a comma after the last item of a multiline array. (https://rubystyle.guide#no-trailing-array-commas)

File: spec/scenarios/credit_note_spec.rb [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        ]
```
Comment (jdenquin) [inline spec/scenarios/credit_note_spec.rb:102]: Style/TrailingCommaInArguments: Put a comma after the last parameter of a multiline method call. (https://rubystyle.guide#no-trailing-params-comma)

**Pull request merged**